File: app/src/main/java/com/vitaliyhtc/dagger2investigation/presentation/productslist/view/adapter/ProductsListAdapter.java
package com.vitaliyhtc.dagger2investigation.presentation.productslist.view.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.vitaliyhtc.dagger2investigation.R;
import com.vitaliyhtc.dagger2investigation.domain.model.Product;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ProductsListAdapter extends RecyclerView.Adapter<ProductsListAdapter.ProductsListViewHolder> {

    private OnProductClickListener mOnProductClickListener;
    private List<Product> mProducts;

    public ProductsListAdapter() {
        mProducts = new ArrayList<>();
    }

    public void setOnProductClickListener(OnProductClickListener onProductClickListener) {
        mOnProductClickListener = onProductClickListener;
    }

    @Override
    public int getItemCount() {
        return mProducts.size();
    }

    public void appendToProducts(List<Product> products) {
        int c1 = getItemCount();
        mProducts.addAll(products);
        notifyItemRangeInserted(c1, products.size());
    }

    public void appendToProducts(Product product) {
        mProducts.add(product);
        notifyItemInserted(getItemCount() - 1);
    }

    @Override
    public ProductsListViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.list_item_product, parent, false);
        return new ProductsListViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ProductsListViewHolder holder, int position) {
        holder.bind(mProducts.get(position));
    }

    class ProductsListViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.item_title)
        TextView textView;
        @BindView(R.id.image_view_product_small)
        ImageView imageView;

        private Context mContext;

        ProductsListViewHolder(View view) {
            super(view);
            ButterKnife.bind(this, view);
            mContext = view.getContext();
        }

        @OnClick(R.id.rl_main)
        void onClick() {
            mOnProductClickListener.onProductClick(mProducts.get(getAdapterPosition()).getId());
        }

        void bind(Product product) {
            textView.setText(product.getName());
            Picasso.with(mContext)
                    .load(product.getImageThumbUrl())
                    .placeholder(R.drawable.ic_list_item_bg)
                    .error(R.drawable.ic_broken_image)
                    .into(imageView);
        }
    }
}

File: app/src/main/java/com/vitaliyhtc/dagger2investigation/presentation/productslist/view/adapter/ProductsListAdapter.kt
package com.vitaliyhtc.dagger2investigation.presentation.productslist.view.adapter

import android.content.Context
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import com.vitaliyhtc.dagger2investigation.R
import com.vitaliyhtc.dagger2investigation.domain.model.Product
import com.vitaliyhtc.dagger2investigation.utils.layoutInflater
import kotlinx.android.synthetic.main.list_item_product.view.*

class ProductsListAdapter : RecyclerView.Adapter<ProductsListAdapter.ProductsListViewHolder>() {

    private val mProducts: ArrayList<Product> = ArrayList()
    //private var mOnProductClickListener: OnProductClickListener? = null

    private var mOnProductClickListener: ((id: Int) -> Unit) = {}

    fun setOnProductClickListener(onProductClickListener: (id: Int) -> Unit) {
        mOnProductClickListener = onProductClickListener
    }

    override fun getItemCount(): Int {
        return mProducts.size
    }

    fun appendToProducts(products: List<Product>) {
        val positionStart = itemCount
        mProducts.addAll(products)
        notifyItemRangeInserted(positionStart, products.size)
    }

    fun appendToProducts(product: Product) {
        mProducts.add(product)
        notifyItemInserted(itemCount - 1)
    }

    fun updateProducts(products: List<Product>) {
        mProducts.clear()
        mProducts.addAll(products)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProductsListViewHolder {
        val v: View = inflateLayout(parent, R.layout.list_item_product)
        return ProductsListViewHolder(v)
    }

    private fun inflateLayout(parent: ViewGroup, @LayoutRes layoutId: Int): View {
        return parent.context.layoutInflater().inflate(layoutId, parent, false)
    }

    override fun onBindViewHolder(holder: ProductsListViewHolder, position: Int) {
        holder.bind(mProducts[position])
    }

    inner class ProductsListViewHolder(view: View) : RecyclerView.ViewHolder(view) {

        private val mContext: Context = view.context
        private val title = view.item_title
        private val image = view.image_view_product_small
        private val imageFavorite = view.iv_favorite

        init {
            view.setOnClickListener { mOnProductClickListener(mProducts[adapterPosition].id) }
        }

        fun bind(product: Product) {
            title.text = product.name
            Picasso.with(mContext)
                    .load(product.image_thumb_url)
                    .placeholder(R.drawable.ic_list_item_bg)
                    .error(R.drawable.ic_broken_image)
                    .into(image)
            if (product.is_favorite) {
                imageFavorite.setImageDrawable(mContext.resources.getDrawable(R.drawable.ic_star_black_24dp))
            } else {
                imageFavorite.setImageDrawable(mContext.resources.getDrawable(R.drawable.ic_star_border_black_24dp))
            }
        }
    }

}